ctr-nonce-reuse/src/ctr_game.rs

```diff
@@ -0,0 +1,196 @@
+        }
+    }
+
+    pub fn restart(&mut self) {
+        self.rng = rand::thread_rng();
+
+        self.key = self.rng.gen();
+        self.nonce = self.rng.gen();
+        //self.aes_ctr = Aes128Ctr64LE::new(&key.into(), &nonce.into());
+
+        let flag_str = self.create_random_flag();
+        self.flag_plaintext = flag_str.into();
+
+        self.create_ciphertexts(EAVESDROPPED_MSG_COUNT);
+    }
+
+    fn create_random_flag(&mut self) -> String {
+        let left_random_bytes: [u8; 4] = self.rng.gen();
+        let right_random_bytes: [u8; 4] = self.rng.gen();
+        let mut flag_str = String::new();
+
+        write!(
+            flag_str,
+            "FLAG{{{}_CtR_1sNT_SeCC_{}}}",
+            Self::byte_to_hex(&left_random_bytes),
+            Self::byte_to_hex(&right_random_bytes)
+        )
+        .expect("String formatting can not fail");
+
+        flag_str
+    }
+
+    fn create_ciphertexts(&mut self, count: usize) {
+        self.ciphertexts_vec.clear();
+        let mut chosen_plaintext: Vec<_> = self
+            .plaintexts_vec
+            .choose_multiple(&mut self.rng, count)
+            .collect();
+        self.known_plaintext = (*chosen_plaintext.choose(&mut self.rng).expect("a")).clone();
+        chosen_plaintext.push(&self.flag_plaintext);
+        chosen_plaintext.shuffle(&mut self.rng);
+        for plaintext in chosen_plaintext {
+            let ciphertext = self.encrypt_bytes(plaintext);
+            self.ciphertexts_vec.push(ciphertext);
+        }
+    }
+
+    pub fn get_hex_ciphertexts(&self) -> Vec<String> {
+        let mut result = Vec::new();
+
+        for ciphertext in &self.ciphertexts_vec {
+            result.push(Self::byte_to_hex(ciphertext));
+        }
+
+        result
+    }
+
+    pub fn get_known_plaintext(&self) -> String {
+        String::from_utf8_lossy(&self.known_plaintext).to_string()
+    }
+
+    pub fn submit_flag(&self, flag_request: String) -> bool {
+        flag_request.into_bytes().eq(&self.flag_plaintext)
+    }
+
+    fn encrypt_bytes(&self, plaintext_vec: &[u8]) -> Vec<u8> {
+        let mut result = plaintext_vec.to_owned();
+
+        let mut cipher = Aes128Ctr64LE::new(&self.key.into(), &self.nonce.into());
+        cipher.apply_keystream(&mut result);
+
+        result
+    }
+
+    /*
+    fn decrypt_bytes(&self, ciphertext_vec: &[u8]) -> Vec<u8> {
+        self.encrypt_bytes(ciphertext_vec)
+    }
+    */
+
+    pub fn byte_to_hex(data: &[u8]) -> String {
+        hex::encode(data)
+    }
+
+    pub fn hex_to_text(data: String) -> String {
+        let decoded = hex::decode(data);
+        if decoded.is_err() {
+            return String::from("[unprintable]");
+        }
+        let decoded = decoded.expect("We have already handled the error");
+
+        let decoded = String::from_utf8(decoded);
+        if decoded.is_err() {
+            return String::from("[unprintable]");
+        }
+        decoded.expect("We have already handled the error")
```

Comment:
There is a pattern in this code, where you first check for an error and then unwrap the result if it's ok. This is how the Result-like type would look in the language without pattern matching. But the Rust is not such language :)
For that operation you could use either `match` or `if let`:
```rust
let decoded = match decoded {
    Ok(decoded) => decoded,
    Err(_) => String::from("[unprintable]")
};
```

```rust
let decoded = if let Ok(decoded) {
    decoded
} else {
    String::from("[unprintable]")
}
```

Or just use the `Result::unwrap_or_else`:
```rust
let decoded = decoded.unwrap_or_else(|| String::from("[unprintable]"));
```

P.S. this is not the only place with this pattern.

---

But in this case, you can use helpers in the `Result` and `Option` structs For example:
 ```suggestion
        hex::decode(data)
            .ok()
            .and_then(|data| String::from_utf8(data).ok())
            .unwrap_or_else(|| String::from("[unprintable]"))
```

What each function does:
- [Result::ok](https://doc.rust-lang.org/std/result/enum.Result.html#method.ok):
  Converts `Ok(data)` into `Some(data)`, and `Err(err)` into `None`.
- [Option::and_then](https://doc.rust-lang.org/std/option/enum.Option.html#method.and_then)
  > Returns [None](https://doc.rust-lang.org/std/option/enum.Option.html#variant.None) if the option is [None](https://doc.rust-lang.org/std/option/enum.Option.html#variant.None), otherwise calls f with the wrapped value and returns the result.
- [Option::unwrap_or_else](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else):
  > Returns the contained [Some](https://doc.rust-lang.org/std/option/enum.Option.html#variant.Some) value or computes it from a closure.
  
  Note that we could use just the [Option::unwrap_or](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or), but it would create a new string in any case (even if the result is ok), which makes an allocation. And we could avoid it with the `unwrap_or_else` which will make the allocation only if the result is None.